src/Features/Protocol/LanguageServices/RoslynLanguageService.cs

```diff
@@ -0,0 +1,673 @@
+
+        // TODO - Move hierarchicalDocumentSymbolSupport to client capabilities.
+        // https://github.com/dotnet/roslyn/projects/45#card-20033973
+        private readonly bool _hierarchicalDocumentSymbolSupport;
+        private readonly ClientCapabilities _clientCapabilities;
+
+        public RoslynLanguageService(ClientCapabilities clientCapabilities, bool hierarchicalDocumentSymbolSupport)
+        {
+            _hierarchicalDocumentSymbolSupport = hierarchicalDocumentSymbolSupport;
+            _clientCapabilities = clientCapabilities;
+        }
+
+        public ServerCapabilities GetServerCapabilities()
+        {
+            return new ServerCapabilities
+            {
+                DefinitionProvider = true,
+                ReferencesProvider = true,
+                DocumentHighlightProvider = true,
+                DocumentSymbolProvider = true,
+                WorkspaceSymbolProvider = true,
+                HoverProvider = true,
+            };
+        }
+
+        /// <summary>
+        /// Answers a document symbols request by returning a list of symbols in the document.
+        /// https://microsoft.github.io/language-server-protocol/specification#textDocument_documentSymbol
+        /// </summary>
+        /// <param name="solution">the solution containing the document.</param>
+        /// <param name="request">the document to get symbols from.</param>
+        /// <param name="cancellationToken">a cancellation token.</param>
+        /// <returns>a list of symbols in the document.</returns>
+        public async Task<object[]> GetDocumentSymbolsAsync(Solution solution, DocumentSymbolParams request, CancellationToken cancellationToken)
+        {
+            var document = solution.GetDocument(request.TextDocument.Uri);
+            if (document == null)
+            {
+                return Array.Empty<SymbolInformation>();
+            }
+
+            var symbols = ArrayBuilder<object>.GetInstance();
+
+            var navBarService = document.Project.LanguageServices.GetService<INavigationBarItemService>();
+            var navBarItems = await navBarService.GetItemsAsync(document, cancellationToken).ConfigureAwait(false);
+            if (navBarItems.Count == 0)
+            {
+                return symbols.ToArrayAndFree();
+            }
+
+            var compilation = await document.Project.GetCompilationAsync(cancellationToken).ConfigureAwait(false);
+            var tree = await document.GetSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
+            var text = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+
+            // TODO - Return more than 2 levels of symbols.
+            // https://github.com/dotnet/roslyn/projects/45#card-20033869
+            return await CreateDocumentSymbolOrSymbolInformations().ConfigureAwait(false);
+
+            // local functions
+            async Task<object[]> CreateDocumentSymbolOrSymbolInformations()
```

Comment:
can you make these static?  it's then clear exactly what they capture and manipulate.